tests/gnrc_sock_neterr/main.c

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (C) 2019 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     tests
+ * @{
+ *
+ * @file
+ *
+ * @author      Martine Lenders <m.lenders@fu-berlin.de>
+ *
+ * @}
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include "msg.h"
+#include "net/sock/udp.h"
+
+#define TEST_PORT               (38664U)
+#define TEST_LINK_LOCAL_REMOTE  { 0xfe, 0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, \
+                                  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02 }
+#define TEST_GLOBAL_REMOTE      { 0x20, 0x01, 0x0d, 0xb8, 0x00, 0x00, 0x00, 0x00, \
+                                  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02 }
+#define TEST_PAYLOAD            { 0x01, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef }
+
+static const uint8_t _test_link_local_remote[] = TEST_LINK_LOCAL_REMOTE;
+static const uint8_t _test_global_remote[] = TEST_GLOBAL_REMOTE;
+static const uint8_t _test_payload[] = TEST_PAYLOAD;
+
+static msg_t _msg_queue[8U];
+static sock_udp_t _udp_sock;
+
+int _test_udp_send(sock_udp_ep_t *remote, char *errno_str, int exp_res) {
+
+    int res = sock_udp_send(&_udp_sock, _test_payload, sizeof(_test_payload),
+                        remote);
+    /* remote is not reachable or route does not exist, so it should return an
+     * error */
+    if (-res == exp_res) {
+        printf("SUCCESS: error code %s (%li == %i)\n",
+               errno_str, (long)(-res), exp_res);
+        return 0;
+    }
+    else {
+        printf("FAILURE: gnrc_udp_send() had an unexpected error code: %li\n",
+               (long int)res);
+        return 1;
+    }
+}
+
+int main(void)
+{
+    sock_udp_ep_t local = SOCK_IPV6_EP_ANY;
+    sock_udp_ep_t remote = SOCK_IPV6_EP_ANY;
+
+    local.port = TEST_PORT;
+    remote.port = TEST_PORT - 1;
+
+    msg_init_queue(_msg_queue, ARRAY_SIZE(_msg_queue));
+    sock_udp_create(&_udp_sock, &local, NULL, 0);
+
+    memcpy(remote.addr.ipv6, _test_link_local_remote,
+           sizeof(_test_link_local_remote));
+    _test_udp_send(&remote, "EHOSTUNREACH", EHOSTUNREACH);
```

Comment:
Maybe add something like

```suggestion
#ifdef MODULE_GNRC_SIXLOWPAN
    _test_udp_send(&remote, "OK", -sizeof(_test_payload));
#else
    _test_udp_send(&remote, "EHOSTUNREACH", EHOSTUNREACH);
#endif
```

Reply:
I don't expect anyone to test this this way in the future.

Reply:
I view the code in `examples/` and `tests/` as extended documentation. I suppose I'm not the only one who does that.

So when I want to understand how to use the `gnrc_neterr` module, I would look at `tests/gnrc_sock_neterr` and conclude that sending a datagram to a non-existing link-local address would always generate an error - after all, there is a testcase to ensure it is so.
It wouldn't be an unreasonable thing to assume if `ACK_REQ` is enabled.

So at least a comment would be appropriate that with 6LoWPAN the behaviour is different, referencing RFC 6775.

Reply:
> So when I want to understand how to use the `gnrc_neterr` module, I would look at `tests/gnrc_sock_neterr` and conclude that sending a datagram to a non-existing link-local address would always generate an error - after all, there is a testcase to ensure it is so.

Again, if you change the test parameters (adding a device that acts differently under IPv6) you may get different results.

> It wouldn't be an unreasonable thing to assume if `ACK_REQ` is enabled.

I don't get this comment... what has an option in the MAC layer to do with error handling in the application layer? The `sock` API is not there to ensure the packet reached the next host. This is also not how sockets in POSIX work. If you want reliability there you have to configure the socket as such (e.g. by using TCP).

I will however adapt the test.

Reply:
> I will however ~adapt the test.~ add a comment.

The `#ifdef MODULE_GNRC_SIXLOWPAN` is the wrong macro. It should rather check if the netif the packet is sent over is a 6LN. This however is getting too complicated for what is supposed to be a simple test, just for the case that somebody gets the idea of changing test parameters around (btw if `GNRC_NETIF_NUMOF>1` the behavior for link-local addresses changes yet again!) and expecting the same results.

Reply:
> I don't get this comment... what has an option in the MAC layer to do with error handling in the application layer? The sock API is not there to ensure the packet reached the next host. This is also not how sockets in POSIX work. 

I'm just saying that someone not familiar with the implementation might assume the lower layer only reports success then if a link-layer ACK was received.
But the comment should help clear that up.